### PR TITLE
LxMERT Integration with DroneVQA Interface

### PR DESCRIPTION
**Pull request by marcoperic**
### How to run this?

- An LxMERT fine-tuned model has been uploaded to the sharepoint, in a new folder named "lxmert_tuned_models." Please place the .pth file in the same directory as application/utils.py

- To most accurately gauge the performance of the fine-tuned model, be sure to clear your cache (somewhere in C:\Users\\{name}\\.cache; check the console for exact directory and folder) when trying to load the fine-tuned model; otherwise, it will load a cached LxMERT version (unc-nlp/lxmert-vqa-uncased). 

### Possible poor performance:

While unlikely, it is possible that the performance of the fine-tuned VQA model will be dismal, as the architecture of the model does not exactly match the architecture in the config.js file. As a result, not all of the weights are used during inference. My next priority, alongside fine-tuning with a greater amount of data, is to export the exact config.js file so that all of the weights of this fine-tuned model can be used. 

The poor performance _will not_ be an expected part of our training process and should resolve itself once the architecture laid out in config.json matches the architecture of the model. **This may take some time to resolve**. In the meantime, however, fine-tuning and training is still possible.

### Training information
Fine-tuning took place on an RTX 3070 Ti consisting of a tiny generalized dataset of 512 images. 
The process took about 120 seconds to complete.  
9 llayers
5 xlayers
5 rlayers
batch size of 32, learning rate of 5e-5, 10 epochs.

### Further details and memory solutions
The model was already pretrained and can be acquired from: http://nlp.cs.unc.edu/data/model_LXRT.pth
Memory fragmentation was a problem, as the 3070 Ti only has 8GB of VRAM. As such, the following possibilities were explored:
- Reduced batch size
- Offloaded model to CPU
- Different optimizers were used
- Exclusively CPU training (about 20s per epoch with 512 images; certainly doable for small datasets).
- PYTORCH_CUDA_ALLOC_CONF parameter used and set to 324MB to eliminate fragmentation. (solution currently in use)

### Misc. notes:

- I did not test the performance of this pull request. The DroneVQA software is not linking to the simulation environment on my machine, so I created this pull request while I trouble shoot it, so that you all can test it alongside me and provide feedback.

- GPU acceleration works fine. If CUDA is not available for you, you can map it to CPU when loading the model.

### CODE REVIEW

File: Application/utils.py [hunk 1]
```diff
@@ -83,7 +83,11 @@ def setupLxmertTransformer():
 
     # Define the model
     lxmert_tokenizer = LxmertTokenizer.from_pretrained("unc-nlp/lxmert-base-uncased")
-    lxmert_vqa = LxmertForQuestionAnswering.from_pretrained("unc-nlp/lxmert-vqa-uncased")
+
+    # *** OLD MODEL. Load this if no local model is present.
+    # lxmert_vqa = LxmertForQuestionAnswering.from_pretrained("unc-nlp/lxmert-vqa-uncased")
```
Comment (Perl-R) [inline Application/utils.py]: We want to have the base and the fine tuned model loaded into the GUI. It looks this line is removing the base model. 
You might have to look at adding the finetuned model around here in the code: https://github.com/jenniferolenchak/DroneVQA/blob/main/Application/application.py#L73

**Review by Perl-R — CHANGES_REQUESTED** [review]

File: Application/utils.py (hunk 1 above)
Comment (marcoperic) [reply Application/utils.py]: All set.

File: Application/utils.py [hunk 2]
```diff
@@ -83,15 +83,33 @@ def setupLxmertTransformer():
 
     # Define the model
     lxmert_tokenizer = LxmertTokenizer.from_pretrained("unc-nlp/lxmert-base-uncased")
-    lxmert_vqa = LxmertForQuestionAnswering.from_pretrained("unc-nlp/lxmert-vqa-uncased")
+
+    # *** Original model
+    lxmert_vqa = LxmertForQuestionAnswering.from_pretrained("unc-nlp/lxmert-vqa-uncased") 
+    lxmert_vqa_finetuned = LxmertForQuestionAnswering.from_pretrained(pretrained_model_name_or_path='lxmert_best_model.pth', config='config.json')
```
Comment (Perl-R) [inline Application/utils.py]: Can we remove lines 87, 89, and 96?

**Review by Perl-R — COMMENTED** [review]
We also need to add the corresponding button in the GUI screen to use the model. You may have to look around here in the code to add the button: https://github.com/jenniferolenchak/DroneVQA/blob/main/Application/VQAInteractionScreen.py#L188

File: Application/utils.py (hunk 2 above)
Comment (marcoperic) [reply Application/utils.py]: good catch

**Comment by jenniferolenchak** [comment]
**Going forward, please do not request a review for unfinished and unreviewed code.**

To start a pull request, changes should either:

- Have passed at least initial testing

~OR~

- Be in-progress and marked as a draft

**Comment by marcoperic** [comment]
Button was added to GUI. Works when clicked. Uses the fine-tuned LXMERT model. Not super surprisingly, LXMERT fine-tuned model performs terribly. I suspect that it is because of the configuration files, it is not able to use all of the weights and is therefore performing poorly. Will fix

**Review by Perl-R — APPROVED** [review]
Ayden agreed to make some UI fixes in his next upcoming PR implementing a finetuned ViLT. Since he plans to make those changes, I will move forward with approving this PR.
